deploy: support for read affinity options per cluster

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #
- **deploy: support for read affinity options per cluster**
Implemented the capability to include read affinity options 
for individual clusters within the ceph-csi-config ConfigMap. 
This allows users to configure the crush location for each 
cluster separately. The read affinity options specified in 
the ConfigMap will supersede those provided via command 
line arguments.

- **util: moved GetNodeLabels() under internal/util/k8s**
- **util: added read affinity related functions and unit testcases**
- **util: added RunsOnKubernetes() function**

Fixes: #4161 

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

Rakshith-R commented on internal/rbd/nodeserver.go:
```diff
@@ -1397,8 +1406,28 @@ func getDeviceSize(ctx context.Context, devicePath string) (uint64, error) {
+		log.FatalLogMsg(err.Error())
+	}
+
+	crushLocationMap, err := util.GetCrushLocationMap(crushLocationLabels, ns.Driver.NodeID)
```
This function call k8s api, which is an expensive operation, to get node labels.
Performance will severely be impacted if we perform such a call on each nodeStage request. 

Please fetch the node labels when the driver starts and store it.

nixpanic replied:
> Please fetch the node labels when the driver starts and store it.

- Is this a problem for a running Ceph-CSI setup when nodes are added later on?
- What happens if the Crush location map is modified?

Rakshith-R replied:
> > Please fetch the node labels when the driver starts and store it.
> 
> * Is this a problem for a running Ceph-CSI setup when nodes are added later on?

No, We just use this in daemonset nodeplugin pod. 
So, when a new node is added, a daemonset pod starts on it and we'll pick up the labels that time.
There is a requirement currently that when node labels are updated, they'll need to restart nodepugin pods andthe mounted volumes for the changed topology to take affect.

> * What happens if the Crush location map is modified?

If you meant Crush Location map modified by means of node topology label update, it is the same as above.
If you meant changes were done to the configmap to update labels, they'll need to restart the nodeplugin pod too.


I expect these changes to be very rare and possibly untouched after initial creation.

nixpanic replied:
I have no idea how often locations/labels are added to a crush map :man_shrugging: If it is clearly documented that changing labels of a node requires restarting, then thats fine.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.27

iPraveenParihar commented:
@Rakshith-R @Madhu-1 
PR is ready for a review

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -45,8 +45,9 @@ type NodeServer struct {
 	VolumeLocks *util.VolumeLocks
-	// readAffinityMapOptions contains map options to enable read affinity.
-	readAffinityMapOptions string
+	NodeLabels  map[string]string
```
please document this new map with a comment

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -280,14 +283,14 @@ func (ns *NodeServer) populateRbdVol(
 // appendReadAffinityMapOptions appends readAffinityMapOptions to mapOptions
 // if mounter is rbdDefaultMounter and readAffinityMapOptions is not empty.
-func (ns NodeServer) appendReadAffinityMapOptions(rv *rbdVolume) {
+func appendReadAffinityMapOptions(rv *rbdVolume, readAffinityMapOptions string) {
```
this can now be a function of the `rbdVolume` struct:
```go
func (rv *rbdVolume) appendReadAffinityMapOptions(readAffinityMapOptions string)
```
which would be cleaner

nixpanic commented on internal/util/crushlocation.go:
```diff
@@ -26,16 +26,11 @@ import (
 // Expects crushLocationLabels in arg to be in the format "[prefix/]<name>,[prefix/]<name>,...",.
 // Returns map of crush location types with its array of associated values.
-func GetCrushLocationMap(crushLocationLabels, nodeName string) (map[string]string, error) {
+func GetCrushLocationMap(crushLocationLabels string, nodeLabels map[string]string) (map[string]string, error) {
```
the comment is not correct anymore. The nodeLabels are not retrieved from the CO system within this function.

nixpanic commented on internal/util/topology.go:
```diff
@@ -34,7 +34,7 @@ const (
 )
 
-func k8sGetNodeLabels(nodeName string) (map[string]string, error) {
+func K8sGetNodeLabels(nodeName string) (map[string]string, error) {
```
maybe you can move this function to a file under `internal/util/k8s/`?

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -1396,9 +1399,46 @@ func getDeviceSize(ctx context.Context, devicePath string) (uint64, error) {
 
-func (ns *NodeServer) SetReadAffinityMapOptions(crushLocationMap map[string]string) {
+func (ns *NodeServer) SetcmdReadAffinityMapOptions(crushLocationMap map[string]string) {
+	ns.cmdReadAffinityMapOptions = constructReadAffinityMapOption(crushLocationMap)
```
Is there a good reason to keep `constructReadAffinityMapOption()` as a separate function? Why not make it

```go
func (ns *NodeServer) constructReadAffinityMapOption(crushLocationMap map[string]string)
```

or something similar?

iPraveenParihar replied:
Certainly, we can make `cmdReadAffinityMapOptions` a public field. This would allow us to directly use it as follows:

```go
ns.cmdReadAffinityMapOptions = ns.constructReadAffinityMapOption(crushLocationMap)
```
This change would eliminate the need for the `SetcmdReadAffinityMapOptions()` method.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.27

Madhu-1 commented on internal/rbd/driver/driver.go:
```diff
@@ -125,8 +126,13 @@ func (r *Driver) Run(conf *util.Config) {
 			})
 	}
 
+	nodeLabels, err = k8s.GetNodeLabels(conf.NodeID)
+	if err != nil {
+		log.FatalLogMsg(err.Error())
+	}
```
This code will cause problems for non-kubernetes CO. do this check only if the labels are passed?

iPraveenParihar replied:
Will move this under 
```go
if conf.EnableReadAffinity {
  ...
}
```

and will make `--enable-ready-affinity` command line flag to be mandatory for enabling read affinity.

Madhu-1 replied:
i was thinking do we really need to have this flag, if the labels are set assume that you will enable read-affinity like we do currently?

iPraveenParihar replied:
> This code will cause problems for non-kubernetes CO. do this check only if the labels are passed?

@Madhu-1 we can have a new `--container-orchestrator` flag _(default - kubernetes)_. So we will be fetching node labels only when the flag is set to _kubernetes_?

Or this way
> and will make --enable-ready-affinity command line flag to be mandatory for enabling read affinity.

Madhu-1 replied:
i would vote for new `--container-orchestrator` which might be useful to decide when to make kube API calls. @nixpanic @Rakshith-R thought?

Rakshith-R replied:
> i was thinking do we really need to have this flag, if the labels are set assume that you will enable read-affinity like we do currently?

labels maybe set in the configmap instead. 

Let's have the read-affinity flag set in cmdline as a pre-requisite. 
This setting can be overridden by individual cluster id in configmap.

> i would vote for new `--container-orchestrator` which might be useful to decide when to make kube API calls. @nixpanic @Rakshith-R thought?

I think this entails changes at a lot of places and is out of scope for this pr.

> and will make --enable-ready-affinity command line flag to be mandatory for enabling read affinity.

This option seems sufficient for me.

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -1396,9 +1400,42 @@ func getDeviceSize(ctx context.Context, devicePath string) (uint64, error) {
+func (ns *NodeServer) getReadAffinityMapOptionFromConfigMap(clusterID string) string {
+	crushLocationLabels, err := util.GetReadAffinityOptions(util.CsiConfigFile, clusterID)
+	if err != nil {
+		log.FatalLogMsg(err.Error())
```
it should not be Fatal you need to return the error message from the function for all cases

Madhu-1 commented on internal/util/k8s/node.go:
```diff
@@ -0,0 +1,23 @@
+package k8s
```
add missing license code

Madhu-1 commented on internal/util/k8s/node.go:
```diff
@@ -0,0 +1,23 @@
+)
+
+func GetNodeLabels(nodeName string) (map[string]string, error) {
+	client, err := NewK8sClient()
+	if err != nil {
+		return nil, fmt.Errorf("can not get node %q information, failed "+
+			"to connect to Kubernetes: %w", nodeName, err)
+	}
```
can you check is it possible to re-use the client instead of creating one always

Madhu-1 commented on charts/ceph-csi-rbd/templates/nodeplugin-daemonset.yaml:
```diff
@@ -87,8 +87,7 @@ spec:
 {{- if .Values.nodeplugin.profiling.enabled }}
             - "--enableprofiling={{ .Values.nodeplugin.profiling.enabled }}"
 {{- end }}
-            - "--enable-read-affinity={{ and .Values.readAffinity .Values.readAffinity.enabled }}"
```
any specific reason to remove `enable-read-affinity` flag?

Madhu-1 commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -282,7 +287,6 @@ topology:
 # readAffinity:
 # Enable read affinity for RBD volumes. Recommended to
 # set to true if running kernel 5.8 or newer.
-# enabled: false
```
lets not remove it, it can cause breaking change for existing clusters

Madhu-1 commented on cmd/cephcsi.go:
```diff
@@ -81,13 +81,13 @@ func init() {
 		"",
 		"list of Kubernetes node labels, that determines the"+
 			" CRUSH location the node belongs to, separated by ','")
+	flag.StringVar(&conf.ContainerOrchestrator, "container-orchestrator", "kubernetes", "container orchestrator")
```
my suggestion was to add this extra flag but not to remove the `EnableReadAffinity` existing flag :)

Madhu-1 commented on docs/deploy-rbd.md:
```diff
@@ -224,6 +223,12 @@ If enabled, this option will be added to all RBD volumes mapped by Ceph CSI.
 
+Read affinity can be configured for individual clusters within the
+`ceph-csi-config` ConfigMap. This allows configuring the crush location labels
+for each cluster separately. The crush location labels specified in the
```
`for each cluster separately` to `for each ceph cluster separately`

Madhu-1 commented on internal/rbd/driver/driver.go:
```diff
@@ -125,8 +126,13 @@ func (r *Driver) Run(conf *util.Config) {
 
-	if conf.EnableReadAffinity {
-		crushLocationMap, err = util.GetCrushLocationMap(conf.CrushLocationLabels, conf.NodeID)
+	nodeLabels, err = k8s.GetNodeLabels(conf.NodeID)
```
call this function only if `ContainerOrchestrator` is kubernetes

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -1396,9 +1397,12 @@ func getDeviceSize(ctx context.Context, devicePath string) (uint64, error) {
+// ConstructReadAffinityMapOption constructs a read affinity map option based on the provided crushLocationMap.
+// It appends crush location labels in the format
+// "read_from_replica=localize,crush_location=label1:value1|label2:value2|...".
+func ConstructReadAffinityMapOption(crushLocationMap map[string]string) string {
```
If this function can be reused move this to util as we need this for cephfs as well?

iPraveenParihar replied:
Yea, I thought to move this to util during CephFS work.

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -1412,5 +1416,39 @@ func (ns *NodeServer) SetReadAffinityMapOptions(crushLocationMap map[string]stri
+// getReadAffinityMapOptions retrieves the readAffinityMapOptions from the CSI config file if it exists.
+// If not, it falls back to returning the `CLIReadAffinityMapOptions` from the command line.
+// If neither of these options is available, it returns an empty string.
+func (ns *NodeServer) getReadAffinityMapOptions(clusterID string) (string, error) {
```
Move this to until as this can be reused for cephfs asl well?

Madhu-1 commented on internal/util/k8s/node.go:
```diff
@@ -0,0 +1,39 @@
+/*
+Copyright 2020 The CephCSI Authors.
```
2023?

Madhu-1 commented on cmd/cephcsi.go:
```diff
@@ -88,6 +88,8 @@ func init() {
 		"",
 		"list of Kubernetes node labels, that determines the"+
 			" CRUSH location the node belongs to, separated by ','")
+	flag.StringVar(&conf.ContainerOrchestrator, "container-orchestrator", "kubernetes",
```
```suggestion
flag.StringVar(&conf.ContainerOrchestrator, "container-orchestrator", "kubernetes", "container orchestrator where cephcsi is running")
```

Madhu-1 commented on deploy/csi-config-map-sample.yaml:
```diff
@@ -66,6 +66,15 @@ data:
         }
         "nfs": {
           "netNamespaceFilePath": "<kubeletRootPath>/plugins/nfs.csi.ceph.com/net",
+        },
```
please add more details about this configuration in this file comments

Madhu-1 commented on docs/deploy-rbd.md:
```diff
@@ -48,8 +48,7 @@ make image-cephcsi
 | `--skipforceflatten`     | `false`                       | skip image flattening on kernel < 5.2 which support mapping of rbd images which has the deep-flatten feature                                                                                                                                                                         |
 | `--maxsnapshotsonimage`  | `450`                         | Maximum number of snapshots allowed on rbd image without flattening                                                                                                                                                                                                                  |
 | `--setmetadata`          | `false`                       | Set metadata on volume                                                                                                                                                                                                                                                               |
-| `--enable-read-affinity` | `false`                       | enable read affinity                                                                                                                                                                                                                                                                 |
```
This required to be set?

Madhu-1 commented on docs/deploy-rbd.md:
```diff
@@ -215,7 +214,7 @@ for more details.
 
 This can be enabled by adding labels to Kubernetes nodes like
 `"topology.io/region=east"` and `"topology.io/zone=east-zone1"` and
-passing command line arguments `"--enable-read-affinity=true"` and
```
Need to revert this back?

Madhu-1 commented on internal/util/k8s/client.go:
```diff
@@ -46,5 +52,9 @@ func NewK8sClient() (*kubernetes.Clientset, error) {
 		return nil, fmt.Errorf("failed to create client: %w", err)
 	}
 
+	if kubeclient == nil {
```
This check is not required here?

Madhu-1 commented on internal/util/read_affinity.go:
```diff
@@ -0,0 +1,82 @@
+	"github.com/ceph/ceph-csi/internal/util/log"
+)
+
+// ConstructReadAffinityMapOption constructs a read affinity map option based on the provided crushLocationMap.
+// It appends crush location labels in the format
+// "read_from_replica=localize,crush_location=label1:value1|label2:value2|...".
+func ConstructReadAffinityMapOption(crushLocationMap map[string]string) string {
+	if len(crushLocationMap) == 0 {
+		return ""
+	}
+
+	var b strings.Builder
+	b.WriteString("read_from_replica=localize,crush_location=")
+	first := true
+	for key, val := range crushLocationMap {
+		if first {
+			b.WriteString(fmt.Sprintf("%s:%s", key, val))
+			first = false
+		} else {
+			b.WriteString(fmt.Sprintf("|%s:%s", key, val))
+		}
+	}
+
+	return b.String()
+}
+
+// GetReadAffinityMapOptions retrieves the readAffinityMapOptions from the CSI config file if it exists.
+// If not, it falls back to returning the `cliReadAffinityMapOptions` from the command line.
+// If neither of these options is available, it returns an empty string.
+func GetReadAffinityMapOptions(
```
Add unit testing for this one?

Madhu-1 commented on internal/util/read_affinity.go:
```diff
@@ -0,0 +1,82 @@
+
+	crushLocationMap, err := GetCrushLocationMap(configCrushLocationLabels, nodeLabels)
+	if err != nil {
+		log.FatalLogMsg(err.Error())
```
log as error and a return err is missing here?

Madhu-1 reviewed (commented):
If possible add a test case where crush rules are set in configmap as well.

Madhu-1 reviewed (dismissed):
LGTM

nixpanic commented on internal/util/read_affinity.go:
```diff
@@ -0,0 +1,76 @@
+// /*
```
nit: the commenting looks very weird. can you use the same as what is done in other files?

nixpanic commented on cmd/cephcsi.go:
```diff
@@ -88,6 +88,8 @@ func init() {
 		"list of Kubernetes node labels, that determines the"+
 			" CRUSH location the node belongs to, separated by ','")
+	flag.StringVar(&conf.ContainerOrchestrator, "container-orchestrator", "kubernetes",
+		"container orchestrator where cephcsi is running")
```
It would be better to auto-detect this. You can check for the `KUBERNETES_SERVICE_HOST` environment variable, for example. When an other CO is supported, detection for that can be added.

A `config.GetContainerOrchestrator()` function can take care of the detection, and return a string constant for Kubernetes (in the `internal/util` package):

```go
type ContainerOrchestrator string

const (
    ContainerOrchestratorKubernetes = ContainerOrchestrator("kubernetes")
)

func (c *Config) GetContainerOrchestrator() ContainerOrchestrator {
    // do detection
    return ContainerOrchestratorKubernetes
}
```

iPraveenParihar replied:
This will be out of scope for this PR and we can take up this in another PR.

nixpanic replied:
I think it would be better to do it in this PR, so that there is no need to add the `container-orchestrator` parameter, and remove it afterwards (breaks backwards compatibility).

iPraveenParihar replied:
What about the default value for the flag? 
I was considering setting the flag's default value to an empty string. This way, if the user doesn't pass a value for the flag, we can automatically detect it. Does that sound good?

nixpanic replied:
Ideally users (or other deployment systems like Rook) do not need to pass a flag at all, so auto-detection should be the default.

If auto-detection is the default, then there is no good reason to have the flag anymore :smiley:

nixpanic replied:
Maybe add a `func RunsOnKubernetes() bool` to the `internal/util/k8s` package?

nixpanic reviewed (commented):
Some minor changes to improve things. It would be nice to split the commit in several ones, one commit for a particular functionality addition, structure them like:

1. moving of code, from one package to an other (like `GetNodeLabels()`)
2. new util functions
3. using the new util functions
4. deployment changes
5. e2e tests (but that may be hard for this PR)

iPraveenParihar committed:
util: moved GetNodeLabels() under internal/util/k8s

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
util: added read affinity related functions and unit testcases

This commit adds util functions related to read affinity
and unit testcases for the same.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
deploy: support for read affinity options per cluster

Implemented the capability to include read affinity options
for individual clusters within the ceph-csi-config ConfigMap.
This allows users to configure the crush location for each
cluster separately. The read affinity options specified in
the ConfigMap will supersede those provided via command line arguments.

Signed-off-by: Praveen M <m.praveen@ibm.com>

nixpanic commented on internal/rbd/driver/driver.go:
```diff
@@ -125,13 +126,17 @@ func (r *Driver) Run(conf *util.Config) {
 
-	if conf.EnableReadAffinity {
-		crushLocationMap, err = util.GetCrushLocationMap(conf.CrushLocationLabels, conf.NodeID)
+	if conf.ContainerOrchestrator == "kubernetes" {
```
`"kubernetes" ` should be a constant, and if set to something else, a warning needs to be emitted.

nixpanic replied:
if `ContainerOrchestrator` is only used here, consider replacing it with `internal/util/k8s.RunsOnKubernetes()` or similar

nixpanic commented on internal/util/csiconfig.go:
```diff
@@ -209,3 +214,20 @@ func GetNFSNetNamespaceFilePath(pathToConfig, clusterID string) (string, error)
+
+// GetCrushLocationLabels returns the crushLocationLabels from csi config for the given clusterID
+// If `readAffinity.enabled` is set to true, else returns an empty string.
+func GetCrushLocationLabels(pathToConfig, clusterID string) (bool, string, error) {
```
Please document the `bool` return value as well.

nixpanic commented on internal/util/csiconfig_test.go:
```diff
@@ -365,3 +365,101 @@ func TestGetNFSNetNamespaceFilePath(t *testing.T) {
+		tc := tt
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			_, got, err := GetCrushLocationLabels(tmpConfPath, tc.clusterID)
```
if there is no need to check the returned bool, don't return it at all. Or validate here what you're expecting.

iPraveenParihar replied:
returned bool value is required and validated in the calling function - `GetReadAffinityMapOptions()` under `internal/util/read_affinity.go`. 

```go
configReadAffinityEnabled, configCrushLocationLabels, err = GetCrushLocationLabels(CsiConfigFile, clusterID)
if err != nil {
  return "", err
}

if !configReadAffinityEnabled {
  return "", nil
}

if configCrushLocationLabels == "" {
  return cliReadAffinityMapOptions, nil
}
```

nixpanic replied:
great, that is an excellent reason to validate the returned bool here too!

nixpanic replied:
I still think it would be good to verify the returned bool here.

iPraveenParihar replied:
Done.

iPraveenParihar committed:
util: added RunsOnKubernetes() function

Signed-off-by: Praveen M <m.praveen@ibm.com>

nixpanic commented on internal/util/k8s/client.go:
```diff
@@ -56,3 +56,11 @@ func NewK8sClient() (*kubernetes.Clientset, error) {
+
+// RunsOnKubernetes checks if the application is running within a Kubernetes cluster
+// by inspecting the presence of the KUBERNETES_SERVICE_HOST environment variable.
+func RunsOnKubernetes() bool {
```
yay, auto-detection!

Can you move this commit before the one that calls this function? It should be possible to at least build the ceph-csi executable after each commit (otherwise `git bisect` breaks).

iPraveenParihar replied:
Done

nixpanic reviewed (dismissed):
Thanks for all the corrections and improvements!

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio queue

nixpanic commented:
It seems the **verify readAffinity support** e2e tests fails.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.26

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.26

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.26

iPraveenParihar commented:
> It seems the **verify readAffinity support** e2e tests fails.

Had forgotten to update the csi config in e2e, made the changes.
PTAL.

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.26

iPraveenParihar commented on PendingReleaseNotes.md:
```diff
@@ -3,3 +3,8 @@
 
 ## Features
+
+RBD
```
@Madhu-1,
added to PendingReleaseNotes.md

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
@mergifyio queue

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.26

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.26

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.26

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio queue

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio queue

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio queue
